### PR TITLE
Test node replacement proactive with vSphere UPI

### PR DESCRIPTION
**Pull request by yitzhak12**

### CODE REVIEW

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 1]
```diff
@@ -72,15 +71,13 @@ def test_nodereplacement_proactive(self, pvc_factory, pod_factory, dc_pod_factor
                     f"results of this test run are all invalid.")
 
         elif config.ENV_DATA['platform'].lower() == constants.VSPHERE_PLATFORM:
-            pytest.skip("Skipping add node in Vmware platform due to "
-                        "https://bugzilla.redhat.com/show_bug.cgi?id=1844521"
-                        )
```
Comment (prsurve) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:77]: we need this bz decorator if ocp version is greater then 4.5

File: conf/ocsci/vsphere_scale_cluster_rhos.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  scale_up: True
+  nodes_scaleup_count: '3'
+  platform: 'vsphere'
+  deployment_type: 'upi'
+  worker_replicas: 3
+  master_replicas: 3
+  worker_num_cpus: '16'
+  master_num_cpus: '4'
+  master_memory: '16384'
+  compute_memory: '65536'
```
Comment (ebenahar) [inline conf/ocsci/vsphere_scale_cluster_rhos.yaml:15]: can we reduce this to 48GB?

Comment (yitzhak12) [reply conf/ocsci/vsphere_scale_cluster_rhos.yaml:15]: Yes sure. Do you think it will be better here?

File: ocs_ci/ocs/node.py [hunk 3]
```diff
@@ -827,3 +827,61 @@ def get_node_az(node):
...
+
+    if not use_existing_node:
+        log.info("Preparing to create a new node...")
+        add_new_node_and_label_upi(node_type, 1)
+    else:
+        node_not_in_ocs = get_worker_nodes_not_in_ocs()[0]
+        ocs_labeled_nodes([node_not_in_ocs], node_type)
+
+
+def ocs_labeled_nodes(nodes, node_type):
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py:866]: please add a docstring.
Also, this function name is not action-oriented. Maybe change to something like `label_node_and_set_selinux_permissions()`

File: ocs_ci/ocs/node.py [hunk 4]
```diff
@@ -827,3 +827,61 @@ def get_node_az(node):
...
+            resource_name=new_node_to_label.name,
+            label=constants.OPERATOR_NODE_LABEL
+        )
+        logging.info(
+            f"Successfully labeled {new_node_to_label.name} "
+            f"with OCS storage label"
+        )
+
+
+def get_worker_nodes_not_in_ocs():
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py:882]: docstring is missing

File: ocs_ci/ocs/node.py [hunk 5]
```diff
@@ -827,3 +827,61 @@ def get_node_az(node):
...
+    ocs_nodes = get_ocs_nodes()
+    ocs_node_names = [n.name for n in ocs_nodes]
+    worker_nodes = get_typed_nodes(constants.WORKER_MACHINE)
+    worker_nodes_not_in_ocs = [n for n in worker_nodes if n.name not in ocs_node_names]
+    return worker_nodes_not_in_ocs
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py:887]:
```suggestion
   return [n for n in worker_nodes if n.name not in ocs_node_names]
```

File: ocs_ci/ocs/node.py (hunk 3 above)
Comment (yitzhak12) [reply ocs_ci/ocs/node.py:866]: Or maybe we should delete `set_selinux_permissions` from the function?

Comment (ebenahar) [reply ocs_ci/ocs/node.py:866]: you mean take the set_selinux_permissions part out to a separate function? If so, I agree

File: ocs_ci/ocs/node.py (hunk 5 above)
Comment (yitzhak12) [reply ocs_ci/ocs/node.py]: Changed

File: ocs_ci/ocs/node.py (hunk 3 above)
Comment (yitzhak12) [reply ocs_ci/ocs/node.py]: Fixed

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 1 above)
Comment (petr-balogh) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:76]: What about this issue which @prsurve pointed out?

**Comment by petr-balogh** [comment]
Last two commits are not signed ;) so please do or squash the commits into one which will be signed otherwise DCO is complaining and prevent to merge.

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 1 above)
Comment (yitzhak12) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:76]: Added an option that skips the test with vSphere if the OCP version is greater then 4.5

Comment (sidhant-agrawal) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:76]: @yitzhak12 Is the  bug 1844521 still reproducible in DC-CP?
I suggest to reconfirm once as this DC is better than previous one.
Also can you please update the bug if you are able to repro the issue.

Comment (yitzhak12) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:76]: Currently, this PR is for deleting an OCS node and replace it with another existing node. I didn't face any issues with that process and also for deploying a cluster with 6 worker nodes on DC-CP. So in the meantime, I will delete the BZ decorator and if the bug will be reproduced will add it again.

**Review by ebenahar — DISMISSED** [review]

**Review by ebenahar — APPROVED** [review]

**Review by ebenahar — DISMISSED** [review]

File: ocs_ci/ocs/node.py [hunk 6]
```diff
@@ -827,3 +827,77 @@ def get_node_az(node):
...
+    Unschedule, drain and delete osd node, and creating a new osd node.
+    At the end of the function there should be the same number of osd nodes as
+    it was in the beginning, and also ceph health should be OK.
+    This function is for vSphere UPI.
+
+    Args:
+        osd_node_name (str): the name of the osd node
+        use_existing_node (bool): If False, create a new node and label it.
+            If True, use an existing node to replace the deleted node
+            and label it.
```
Comment (petr-balogh) [inline ocs_ci/ocs/node.py:845]: Please add one more empty line after this section

File: ocs_ci/ocs/node.py [hunk 7]
```diff
@@ -827,3 +827,77 @@ def get_node_az(node):
...
+
+
+def label_nodes(nodes, label=constants.OPERATOR_NODE_LABEL):
+    """
+    Label nodes
+
+    Args:
+        nodes (list): list of node objects need to label
+        label (str): New label to be assigned for these nodes.
+            Default value is the OCS label
```
Comment (petr-balogh) [inline ocs_ci/ocs/node.py:879]: Please add one more empty line after this section

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 8]
```diff
@@ -8,23 +8,78 @@
...
+        if config.ENV_DATA['deployment_type'] == 'ipi':
+            log.info("Perform delete and create osd node in AWS IPI")
+            node.delete_and_create_osd_node_aws_ipi(osd_node_name)
+
+        elif config.ENV_DATA['deployment_type'] == 'upi':
+            log.info("Perform delete and create osd node in AWS UPI")
+            node.delete_and_create_osd_node_aws_upi(osd_node_name)
+        else:
+            pytest.fail(
+                f"ocs-ci config 'deployment_type' value '{config.ENV_DATA['deployment_type']}' is not valid, "
```
Comment (petr-balogh) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:54]: long line, can you split it better please?

**Review by petr-balogh — COMMENTED** [review]
Looks OK. Once you will fix the comments, ping me and I will approve.

**Review by ebenahar — APPROVED** [review]

**Review by petr-balogh — APPROVED** [review]

**Pull request merged**